src/app/core/internal/impl/ComponentInfo.ts

```diff
@@ -0,0 +1,112 @@
+import { Schema } from "core/schema";
+
+export interface PortConfig {
+    counts: number[];
+};
+
+// Describes legal component configurations.
+// NOTE: The configuration wherein each port group has zero ports is IMPLICITLY LEGAL.
+export interface ComponentInfo {
+    readonly kind: string;
+    readonly portGroups: string[];
+    readonly defaultPortConfig: PortConfig;
+
+    isValidPortConfig(p: PortConfig): boolean;
+    makePortsForConfig(id: Schema.GUID, p: PortConfig): Schema.Port[] | undefined;
+
+    // i.e. Prevents fan-in on digital ports, could prevent illegal self-connectivity
+    isValidPortConnectivity(wires: Map<Schema.Port, Schema.Port[]>): boolean;
+
+    // Also i.e. thumbnail, display name, description, etc.
+}
+
+export interface ComponentInfoProvider {
+    get(kind: string): ComponentInfo | undefined;
+}
+
+// TODO: seems kind of heavy
+export function IsValidPortList(info: ComponentInfo, ports: Schema.Port[]): boolean {
+    let countMap = new Map(info.portGroups.map(g => [g, Number(0)]));
+    for (const p of ports) {
+        let count = countMap.get(p.group);
+        if (count === undefined)
+            return false;
+        count++;
+    }
+    let counts: number[] = info.portGroups.map(g => countMap.get(g) ?? -1);
+    return info.isValidPortConfig({ counts: counts });
+}
+
+//
+// Example "Digital" implementation
+//
+
+class BasicGateInfo implements ComponentInfo {
+    private readonly _kind: string;
+
+    // i.e. with thumbnail, display name, description, etc.
+    public constructor(kind: string) {
+        this._kind = kind;
+    }
+
+    public get kind(): string {
+        return this._kind;
+    }
+    public readonly portGroups: string[] = ["IN", "OUT"];
+    public readonly defaultPortConfig: PortConfig = ({ counts: [2, 1] });
+
+    public isValidPortConfig(p: PortConfig): boolean {
+        // At least 2 inputs, exactly 1 output
+        return p.counts[0] >= 2 && p.counts[1] == 1;
```

Comment:
I feel like a PortConfig counts should be a `Record<string, number>` so that it can instead be written as
`p.counts["IN"] >= 2 && p.counts["OUT"] == 1`, which seems a lot more readable